IntegrationTests/UI.IntegrationTests/UITest.cs

```diff
@@ -33,44 +33,23 @@ void HandleApplicationIdle(object sender, EventArgs e)
             Func<T, Task> runTestAsync)
             where T : Form
         {
-            // If form.Dipose was called by the async test task, closing the window would be done in a strange order.
+            Assert.IsEmpty(Application.OpenForms, $"{Application.OpenForms.Count} open form(s) before test");
+
             T form = null;
             try
             {
-                // Start runTestAsync before calling showForm.
-                // The latter might block until the form is closed, especially if using Application.Run(form).
-
+                // Start runTestAsync before calling showForm, since the latter might block until the form is closed.
+                //
                 // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
                 // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
                 // complete.
                 var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
                 {
-                    try
-                    {
-                        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                        await WaitForIdleAsync();
-                        form = Application.OpenForms.OfType<T>().Single();
-                    }
-                    catch (Exception ex)
-                    {
-                        // This exception might not be visible because showForm might not return if the opened form is not closed.
-                        // Log it for the case the "finally" code does not help.
-                        Console.WriteLine($"{nameof(RunForm)}<{typeof(T).FullName}> async test preparation threw {ex.Demystify()}");
-                        throw;
-                    }
-                    finally
-                    {
-                        Application.OpenForms.OfType<Form>().ForEach(f => f.Close());
-                    }
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    await WaitForIdleAsync();
```

Comment:
📝 Neither of these lines can fail.

Reply:
There is no contract that ensures this claim. Remember Murphy's Law.

Reply:
> There is no contract that ensures this claim.

The contract of the first one only allows failure to occur if cancellation is requested:
https://github.com/microsoft/vs-threading/blob/218ab77a6126819c3a5ba256b6ee5f19b987672d/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs#L145-L148

We are passing `CancellationToken.None`, so this can never happen.

If the contract is violated, the code will enter a catch-all block that terminates the entire process, ensuring the failure will not "leak":

https://github.com/microsoft/vs-threading/blob/218ab77a6126819c3a5ba256b6ee5f19b987672d/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs#L969-L974

The contract for `WaitForIdleAsync` boils down to exactly the same contract as `SwitchToMainThreadAsync`.

> Remember Murphy's Law.

This isn't a very practical way to write test code. Remember that the goal here is _not_ to make tests pass when code is broken (includes test code). The goal is one of the following:

1. Avoid test hangs by causing tests to instead fail
2. Ensure that we have the ability to root cause failures

Generally changes need to be motivated by a demonstrated need one of these areas, and proven to achieve the desired outcome for a non-trivial number of cases.

Reply:
I do not have a substantial reason to expect exceptions from these calls.
But as basic principle, I do not trust any called third party function (anymore) -- including / especially complex libraries.
It is far away from being obvious that any exception will be turned into a crash (e.g. any `new` can throw).

> Remember that the goal here is not to make tests pass when code is broken (includes test code). The goal is one of the following:
> 1. Avoid test hangs by causing tests to instead fail
> 2. Ensure that we have the ability to root cause failures

That's exactly the reason why I wrapped it in a `try finally`.
Then if an (unlikely) exception occurs, the test does not hang but fails with this exception.
Without this, the exception is invisible because the test runner kills the test hanging due to the open form.

And I fell into this trap when I instrumented the code with copy & paste debug output.
How many developers will wrap their debug code in `try catch`?